### PR TITLE
[WIP] Add proper support for h3ulcb/m3ulcb/salvator-x-m3n and fix DomU support in general

### PR DESCRIPTION
**Pull request by otyshchenko1**
While aligning with Gen3-test, pick up a few misc patches for uirfs.sh, u-boot-env.txt, etc.

TODO: Update INSTALL.txt to reflect current state.

### CODE REVIEW

**Comment by rshym** [comment]
Very good idea about copying uirfs.sh.
Clear and easy to understand "comment regarding XT_RCAR_EVAPROPRIETARY_DIR usage".
No opinion on other commits.

**Comment by otyshchenko1** [comment]
TODO is eliminated.

File: INSTALL.txt [hunk 1]
```diff
@@ -25,11 +25,9 @@ It also contains different para-virtualized (PV) backends running inside it in o
...
 3. Android based domain named "DomA". It has different PV frontends running and doesn't have any HW assigned except GPU.
 The latter is shared between "DomD" and "DomA" in current setup. 
-
-It worth to mention that IPMMU for GPU is enabled in case of H3 ES3.0 only. For other supported SoCs the PVRKM
-performs IPA to PA translations using a hypercall to hypervisor.
-
-The ARM TF was patched to bring up all available CPU cores.
+- or -
+4. Linux based simple domain named "DomU". It has different PV frontends running and doesn't have any HW assigned except GPU.
+The latter is shared between "DomD" and "DomU" in current setup.
```
Comment (rshym) [inline INSTALL.txt]: Item 4 contradicts line 19 "There are three domains which ...". Need some alignment.

File: INSTALL.txt [hunk 2]
```diff
@@ -61,16 +60,32 @@ workspace_base_dir = /home/workspace_base
...
+XT_GUESTS_BUILD = "domu"
+
+For example (DomA),
+
+XT_GUESTS_INSTALL = "doma"
+XT_GUESTS_BUILD = "doma"
+
+The XT_RCAR_EVAPROPRIETARY_DIR (DomD and DomU), XT_ANDROID_PREBUILDS_DIR (DomA) variables
+depend on whether prebuilt EVA proprietary packages are going to be used.
+
```
Comment (rshym) [inline INSTALL.txt]: This two lines duplicate following line "If you are going to use prebuilt EVA proprietary packages,.."

**Comment by rshym** [comment]
Also as part of last commit "Update documentation to reflect current state" we need to update adnroid's partitions as we have 'vbmeta' and 'misc' now. May be need separate commit after merge of this pull request.

**Comment by otyshchenko1** [comment]
> Also as part of last commit "Update documentation to reflect current state" we need to update adnroid's partitions as we have 'vbmeta' and 'misc' now. May be need separate commit after merge of this pull request.
+1
If you have exact text you could provide, I will be happy to include to that patch.

File: INSTALL.txt (hunk 1 above)
Comment (otyshchenko1) [reply INSTALL.txt]: ok, will rewrite in the following way:
3. Either Android based domain named "DomA" or Linux based domain named "DomU".
Both have different PV frontends running and don't have any HW assigned except GPU.
The latter is shared between "DomD" and "DomA" ("DomU") in current setup.

File: INSTALL.txt (hunk 2 above)
Comment (otyshchenko1) [reply INSTALL.txt]: Will rephrase

**Comment by rshym** [comment]
> > Also as part of last commit "Update documentation to reflect current state" we need to update adnroid's partitions as we have 'vbmeta' and 'misc' now. May be need separate commit after merge of this pull request.
> 
> +1
> If you have exact text you could provide, I will be happy to include to that patch.

See proposed text below.
@arminn take a look please

It has following partitions (sizes are in MiB and subject to change):
system    3148
vendor     270
misc         1
vbmeta       1
metadata    12
userdata  1000

**Comment by rshym** [comment]
> > > Also as part of last commit "Update documentation to reflect current state" we need to update adnroid's partitions as we have 'vbmeta' and 'misc' now. May be need separate commit after merge of this pull request.
> > 
> > 
> > +1
> > If you have exact text you could provide, I will be happy to include to that patch.
> 
> See proposed text below.
> @arminn take a look please
> 
> It has following partitions (sizes are in MiB and subject to change):
> system 3148
> vendor 270
> misc 1
> vbmeta 1
> metadata 12
> userdata 1000

See below reworked text of 'chapter 2'.



2. Product image structure (for system with DomA)

Product image contains following partitions: dom0, domd, domu/doma.
DomA itself contains it own partitions: system, vendor, misc, vbmeta, metadata, userdata.

You can inspect partitions using losetup and lsblk.
See example below for Ubuntu 18, assuming that image has name 'prod-devel.img'.
Pay attention that text after ## is author's comments to console output.

```
$ sudo losetup --find --partscan --show ./prod-devel.img
/dev/loop23

$ lsblk /dev/loop23
NAME       MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
loop23       7:23   0    7G  0 loop 
├─loop23p1 259:0    0  256M  0 loop                 ## dom0
├─loop23p2 259:1    0    2G  0 loop                 ## domd
└─loop23p3 259:2    0  4.3G  0 loop                 ## doma in this case

$ sudo losetup --find --partscan --show /dev/loop23p3
/dev/loop24

$ lsblk /dev/loop24
NAME       MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
loop24       7:24   0  4.3G  0 loop 
├─loop24p1 259:3    0  3.1G  0 loop                 ## system
├─loop24p2 259:4    0  269M  0 loop                 ## vendor
├─loop24p3 259:5    0    1M  0 loop                 ## misc
├─loop24p4 259:6    0    1M  0 loop                 ## vbmeta
├─loop24p5 259:7    0   11M  0 loop                 ## metadata
└─loop24p6 259:8    0 1000M  0 loop                 ## userdata

sudo losetup -d /dev/loop24
sudo losetup -d /dev/loop23
```

**Comment by otyshchenko1** [comment]
> > > > Also as part of last commit "Update documentation to reflect current state" we need to update adnroid's partitions as we have 'vbmeta' and 'misc' now. May be need separate commit after merge of this pull request.
> > > 
> > > 
> > > +1
> > > If you have exact text you could provide, I will be happy to include to that patch.
> > 
> > 
> > See proposed text below.
> > @arminn take a look please
> > It has following partitions (sizes are in MiB and subject to change):
> > system 3148
> > vendor 270
> > misc 1
> > vbmeta 1
> > metadata 12
> > userdata 1000
> 
> See below reworked text of 'chapter 2'.
> 
> 1. Product image structure (for system with DomA)
> 
> Product image contains following partitions: dom0, domd, domu/doma.
> DomA itself contains it own partitions: system, vendor, misc, vbmeta, metadata, userdata.
> 
> You can inspect partitions using losetup and lsblk.
> See example below for Ubuntu 18, assuming that image has name 'prod-devel.img'.
> Pay attention that text after ## is author's comments to console output.
> 
> ```
> $ sudo losetup --find --partscan --show ./prod-devel.img
> /dev/loop23
> 
> $ lsblk /dev/loop23
> NAME       MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
> loop23       7:23   0    7G  0 loop 
> ├─loop23p1 259:0    0  256M  0 loop                 ## dom0
> ├─loop23p2 259:1    0    2G  0 loop                 ## domd
> └─loop23p3 259:2    0  4.3G  0 loop                 ## doma in this case
> 
> $ sudo losetup --find --partscan --show /dev/loop23p3
> /dev/loop24
> 
> $ lsblk /dev/loop24
> NAME       MAJ:MIN RM  SIZE RO TYPE MOUNTPOINT
> loop24       7:24   0  4.3G  0 loop 
> ├─loop24p1 259:3    0  3.1G  0 loop                 ## system
> ├─loop24p2 259:4    0  269M  0 loop                 ## vendor
> ├─loop24p3 259:5    0    1M  0 loop                 ## misc
> ├─loop24p4 259:6    0    1M  0 loop                 ## vbmeta
> ├─loop24p5 259:7    0   11M  0 loop                 ## metadata
> └─loop24p6 259:8    0 1000M  0 loop                 ## userdata
> 
> sudo losetup -d /dev/loop24
> sudo losetup -d /dev/loop23
> ```

Thank you. This wants a separate patch.

**Pull request merged**